src/logsumexp.jl

```diff
@@ -61,36 +61,59 @@ _logsumexp_onepass_reduce(X, ::Base.EltypeUnknown) = reduce(_logsumexp_onepass_o
 ## Reductions for one-pass algorithm: avoid expensive multiplications if numbers are reduced
 
 # reduce two numbers
-function _logsumexp_onepass_op(x1, x2)
-    a = x1 == x2 ? zero(x1 - x2) : -abs(x1 - x2)
-    xmax = x1 > x2 ? oftype(a, x1) : oftype(a, x2)
+function _logsumexp_onepass_op(x1::T, x2::T) where {T<:Number}
+    xmax, a = if x1 == x2
+        # handle `Inf` etc. correctly
+        x2, zero(x1 - x2)
+    elseif real(x1) > real(x2)
+        x1, x2 - x1
+    else
+        x2, x1 - x2
+    end
     r = exp(a)
     return xmax, r
 end
+_logsumexp_onepass_op(x1::Number, x2::Number) = _logsumexp_onepass_op(promote(x1, x2)...)
 
 # reduce a number and a partial sum
-function _logsumexp_onepass_op(x, (xmax, r)::Tuple)
-    a = x == xmax ? zero(x - xmax) : -abs(x - xmax)
-    if x > xmax
-        _xmax = oftype(a, x)
-        _r = (r + one(r)) * exp(a)
+function _logsumexp_onepass_op(x::Number, (xmax, r)::Tuple{<:Number,<:Number})
```

Comment:
Use the short function form (here and below)?

Reply:
You mean just `_logsumexp_onepass_op(x::Number, (xmax, r)::Tuple{...}) = ...`?

Reply:
Yes.